addons/mrp/models/mrp_bom.py

```diff
@@ -82,8 +82,9 @@ def onchange_product_id(self):
     @api.constrains('product_id', 'product_tmpl_id', 'bom_line_ids')
     def _check_bom_lines(self):
         for bom in self:
+            bom_product = bom.product_id or bom.product_tmpl_id
             for bom_line in bom.bom_line_ids:
-                if bom.product_id and bom_line.product_id == bom.product_id:
+                if bom_product == bom_line.product_id or bom_product == bom_line.product_id.product_tmpl_id:
```

Comment:
In case `bom_product` is a `product.template`, you are comparing apple and oranges in:
```py
bom_product == bom_line.product_id
```

Reply:
Yes.  But then it works because an apple is not an orange :)
I have not decided whether it was an elegant way of taking advantage of dynamic typing or an ugly hack.  Maybe if I renamed the variable `bom_product_or_template` ?

Reply:
Hum yeah but we don't want to do that: https://github.com/odoo/odoo/blob/cd9c192fb356cfdaec2a70d81d64c6b7c34aa7c6/odoo/models.py#L5088

Reply:
In fact it passes without warning because they are still both instances of "BaseModel"

```
    def __eq__(self, other):
        """ Test whether two recordsets are equivalent (up to reordering). """
        if not isinstance(other, BaseModel):
            if other:
                filename, lineno = frame_codeinfo(currentframe(), 1)
                _logger.warning("Comparing apples and oranges: %r == %r (%s:%s)",
                                self, other, filename, lineno)
            return NotImplemented
        return self._name == other._name and set(self._ids) == set(other._ids)
```

But I am not attached to the code so if it's not "odooist", or if it's potentially confusing, let's change it.  I can put this instead it's a little more wordy but less needlessly clever.  One difference is it will let pass the case where you have the BOM target a product template and in the BOM components you pick a variant of that template (not sure that is something that makes much sense)

```
                if bom.product_id:
                    same_product = bom.product_id == bom_line.product_id
                else:
                    same_product = bom.product_tmpl_id == bom_line.product_id.product_tmpl_id
                if same_product:
                    raise ValidationError(_("BoM line product %s should not be the same as BoM product.") % bom.display_name)
```

Looks better?

Reply:
Yep, that would look better. `product.template` and `product.product` may be quite confusing. It is even worse if you'd do `bom_product.id == bom_line.product_id.id`.

Reply:
Sounds good.  I updated the code as requested